### DISCUSSION TITLE
What exactly is it?

### DISCUSSION THREAD

**Discussion by Arctomachine** [nitrojs/nitro#2050] [General]
I have read documentation, but it is still unclear to me - what exactly it is and how it is meant to be used? I also searched other resources, but all search results (except for single 8 minute video) ever mentioned it only in context of nuxt and never as independent product.

Is it web server library like express?
Is it complete backend framework like nest and hono?
Is it platform to build my own framework like above mentioned?
Is it something else different?

For whichever it is, are there other resources to read and watch? More like tutorials/guides of practical application, because docs only have description of building blocks, which by themselves hold little value without prior knowledge of how to assemble them.

Can it be used (for backend part) together with vite (for frontend part) to build new framework? Or will they conflict/overlap too much to be easy to work with?

**Comment by TheAlexLichter** [comment]
> Is it web server library like express?

Yes, but modern, supporting for express middleware + deployable "anywhere" (e.g. edge, deno, ...).

> Is it complete backend framework like nest and hono?

My opinion: Nest and hono are two different categories.
Nitro is definitely comparable to hono (feature-wise)
Compared to a "full-blown" backend framework like Nest (including 1st class scheduling, queues, ~~websockets~~, SSE support) there is still a gap to fill.

> Is it platform to build my own framework like above mentioned?

It doesn't have to be but can be, yes.

**Reply by ivanjeremic** [reply]
what about bun.js?

**Reply by vladkrasn** [reply]
@manniL 
> Compared to a "full-blown" backend framework like Nest there is still a gap to fill

Is it a gap that is aimed to be filled? It's a big undertaking. 
I found this project searching for a more "modern" Nest alternative. Nuxt/unjs ecosystem feels like a natural place for such a project to arise, but I am not sure if such front-end focused ecosystem would want to have a full-blown backend framework inside of it. And if people overseeing it are interested at all in such project.

So if I am interested in a tool that competes with Nest, is Nitro a project to keep an eye on, or is its scope smaller? From the docs is certainly seems like a smaller-scope project

**Reply by TheAlexLichter** [reply]
@vladkrasn I talked about my opinion on that in the [latest DejaVue episode](https://share.transistor.fm/s/71c15ba7#t=36m14s) (around 36:14) and think it is matter of time. I think it could be more about creating standalone runtime-agnostic packages (see e.g. unstorage) and "integrating" them but let's see where things go

**Reply by vladkrasn** [reply]
Podcast posted 3 hours ago. Lucky timing!
Will be watching where nitro goes!

**Comment by pi0** [comment]
Hi. Good notes from @manniL adding some more:

Nitro is a **builder** + **server** + development server + meta-framework (extra goodies such as [crossws](https://crossws.unjs.io)). I think perhaps it is not comparable to any other project while parts of it, can be compared.

Niro provides tools, conventions and HTTP layer to make any WebServer and also builds and optimizes it for deployment.

---

Other HTTP frameworks such as express, (also hono ~kinda) are comparable to [unjs/h3](https://h3.unjs.io) which itself can be deployed to many hosting solutions and is part of Nitro.

**Reply by ivanjeremic** [reply]
So if hono would integrate goodies such as `crossws` & `unstorage` it would be comparable to Nitro? Wonder why the Hono guys don't work with you together on the more low level stuff.

**Reply by pi0** [reply]
With runtime goodies + all deployment presets + configured rollup for dev and prod build yes we could compare.

---

For context, We made h3 and Nitro stack LONG before, and not only core is open, we made smaller parts reusable so more of the ecosystem can leverage and collaborate for with least possible opinions. 

For crossws I did my try (https://github.com/honojs/hono/pull/2265) well there are always opinionated reasons to be fair to not reuse. 

Anyway, if you find any collaboration opportunities, please ping me i love nothing more than making them happen @ivanjeremic ❤️
